### ISSUE TITLE
Map Editor/Music Player Unusable + Key Config Issue

### ISSUE THREAD

**Issue by DeWaddle** [SHNecro/ShanghaiJAR#8]
So far the game itself is functional. The player and editor can't be run for whatever reason however. Additionally, the Key Config has a persisting issue from the old version where one input from either a keyboard or controller will be held and autofire on its own. The key can still be changed though. For controllers. with the Xbox 360 one I use, its R2/right trigger, while it changes to L2/left trigger when its actually pressed.

**Comment by SHNecro** [comment]
-- link obsolete --

I think I've handled the BGM/editor crash, added a hacky workaround to embed OpenAL so it doesn't need to be installed.

I've added "stuck-button" handling to KeyConfig, and adjusted the behavior a bit so all buttons have their own check, but this was a very unfortunate time to realize I'd forgotten my bootleg PS3 controller.

Please let me know if these fixes work for you.

**Comment by SHNecro** [comment]
My bad, the attempted fix at on WIP2 actually broke more stuff.
Put a fix into WIP4, hopefully it works better.

I've extended the "stuck-button" checking to the keyconfig, and slightly modified how it works ingame, if a button has been held down since the game/config started, it will need to be released once before it starts being read.
Previously, I tried some sort of "press any button to enable controller" for cases where someone only used keyboard but some leftover virtual controller driver etc. held down Turbo, but it wasn't much of a solution for actual controller users.

If there are still issues, could you check https://gamepad-tester.com/ and let me know what's showing up?

**Comment by DeWaddle** [comment]
Music Player and Map Editor work like a dream now, but the controls are still experiencing issues. Autofire isn't occurring on the key configuration now, but the trigger and D-Pad inputs aren't registering. Additionally, it seems like the D-Pad error is affecting the game itself now, plus the autofire for speed boost is stuck ingame as well when its still mapped to L2. It functions properly when mapped to any of the other, functional buttons however.

Additionally, the updated SlimDX.dll and Common.dll keep the KeyConfig file open even after its been shut down, requiring a forced shutdown from Task Manager.

**Comment by SHNecro** [comment]
Sorry, I'm having trouble reproducing the issues so this might take some time, the virtual controller I'm using for now doesn't seem to have options for D-Pad and I'm not sure what sort of input could be happening with L2 to cause that problem.

For the .dll/process issue I'm not able to reproduce that either, on my machine it all seems to clean up after itself and close out.
I'm pretty much clueless about dlls/exes so this is just pulling out the same solution that worked for another issue I didn't understand, but another user reported their antivirus was causing the game to crash in OpenGL and blocking inputs in DX9.
Seems farfetched, but some of the stuff the game does might raise enough red flags to cause issues, like capturing keypresses, having embedded .dlls, performing encryption/decryption of the resource files, etc.

**Comment by DeWaddle** [comment]
Its probably for the best to leave the controller inputs be for the time being. Reverting the changes to them would probably leave the game in a more stable, playable state than it is now. Weird thing to note: plugging the controller in when KeyConfig's already open will allow inputs to register as they normally would, provided that they're held down to override the forcible R2 (L2) input.

**Comment by SHNecro** [comment]
Sorry, I've just got a thing about leaving things half-working, and pretty much every release someone would have an issue with stuck turbo which this "fix" was supposed to address.
On the plus side, now that I have access to my actual controller again, I see your exact problem and I'm pretty sure I've fixed it for the next release (hopefully soon, unless I get sidetracked again).

Issue turned out to be some "ghost" controller that appeared whenever I plugged in my controller with, as you saw, L2 and R2 constantly pressed. Combined with the my naive handling of multiple controllers, the real controller would "unstick" L2 and R2 since they weren't pressed, then the ghost would keep them pressed down constantly.
